wallet/src/repository/wallet/mod.rs

```diff
@@ -950,153 +1009,166 @@ where
+            let key_balance = state.utxo_set.get(&input.output_pointer().into()).unwrap();
+
+            let model::Path {
+                keychain, index, ..
+            } = self.db.get(&keys::pkh(&key_balance.pkh))?;
+            let parent_key = state
+                .keychains
+                .get(keychain as usize)
+                .expect("could not get keychain");
+
+            let extended_sign_key =
+                parent_key.derive(&self.engine, &types::KeyPath::default().index(index))?;
+
+            let sign_key = extended_sign_key.into();
+
+            let public_key = From::from(types::PK::from_secret_key(&self.engine, &sign_key));
+            let signature = From::from(types::signature::sign(
+                &self.engine,
+                sign_key,
+                sign_data.as_ref(),
+            )?);
+
+            keyed_signatures.push(types::KeyedSignature {
+                signature,
+                public_key,
+            });
+        }
+
+        Ok(keyed_signatures)
+    }
+
     fn create_vt_transaction_components(
         &self,
         state: &mut State,
-        value: u64,
+        outputs: Vec<ValueTransferOutput>,
         fee: u64,
-        recipient: Option<(types::PublicKeyHash, u64)>,
-    ) -> Result<types::TransactionComponents> {
-        self.create_transaction_components(state, value, fee, recipient, false)
+    ) -> Result<(Vec<Input>, Vec<ValueTransferOutput>)> {
+        let utxo_strategy = UtxoSelectionStrategy::Random;
+        let tx_pending_timeout = u64::from(state.epoch_constants.checkpoints_period) * 10;
+        let timestamp = u64::try_from(get_timestamp()).unwrap();
+
+        let (inputs, outputs) = self.build_inputs_outputs_inner_wallet(
+            outputs,
+            None,
+            fee,
+            state,
+            timestamp,
+            tx_pending_timeout,
+            None,
+            utxo_strategy,
+        )?;
+
+        Ok((inputs, outputs))
     }
 
     fn create_dr_transaction_components(
         &self,
         state: &mut State,
-        value: u64,
+        request: DataRequestOutput,
         fee: u64,
-    ) -> Result<types::TransactionComponents> {
-        self.create_transaction_components(state, value, fee, None, true)
+    ) -> Result<(Vec<Input>, Vec<ValueTransferOutput>)> {
+        let utxo_strategy = UtxoSelectionStrategy::Random;
+        let tx_pending_timeout = u64::from(state.epoch_constants.checkpoints_period) * 10;
+        let timestamp = u64::try_from(get_timestamp()).unwrap();
+
+        let (inputs, outputs) = self.build_inputs_outputs_inner_wallet(
+            vec![],
+            Some(&request),
+            fee,
+            state,
+            timestamp,
+            tx_pending_timeout,
+            None,
+            utxo_strategy,
+        )?;
+
+        Ok((inputs, outputs))
     }
 
-    fn create_transaction_components(
+    #[allow(clippy::too_many_arguments)]
+    fn build_inputs_outputs_inner_wallet(
```

Comment:
Comment why this function is slightly different from the one in data_structures. :)